### PR TITLE
KAA-978 Ability to handle device management error codes

### PR DESCRIPTION
**Pull request by Dubland**
/cc @rasendubi @forGGe

### CODE REVIEW

File: client/client-multi/client-c/src/kaa/kaa_bootstrap_manager.c [hunk 1]
```diff
@@ -452,9 +453,10 @@ kaa_error_t kaa_bootstrap_manager_handle_server_sync(kaa_bootstrap_manager_t *se
     return error_code;
 }
 
-kaa_error_t kaa_bootstrap_manager_on_access_point_failed(kaa_bootstrap_manager_t *self
-                                                       , kaa_transport_protocol_id_t *protocol_id
-                                                       , kaa_server_type_t type)
+kaa_error_t kaa_bootstrap_manager_on_access_point_failed(kaa_bootstrap_manager_t *self,
+                                                         kaa_transport_protocol_id_t *protocol_id,
+                                                         kaa_server_type_t type,
+														 uint16_t reason_code)
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_bootstrap_manager.c]: alignment

**Comment by RostakaGmfun** [comment]
Run build

File: client/client-multi/client-c/src/kaa/kaa_bootstrap_manager.c [hunk 2]
```diff
@@ -544,7 +553,7 @@ bool kaa_bootstrap_manager_process_failover(kaa_bootstrap_manager_t *self)
         if (self->next_operations_request_time && current_time >= self->next_operations_request_time) {
             KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "Response bootstrap time expired.");
             kaa_bootstrap_access_points_t * acc_point = (kaa_bootstrap_access_points_t *) kaa_list_get_data(kaa_list_begin(self->bootstrap_access_points));
-            error_code = kaa_bootstrap_manager_on_access_point_failed(self, &acc_point->protocol_id, KAA_SERVER_BOOTSTRAP);
+            error_code = kaa_bootstrap_manager_on_access_point_failed(self, &acc_point->protocol_id, KAA_SERVER_BOOTSTRAP, 0);
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_bootstrap_manager.c]: `0` should be `KAA_BOOTSTRAP_SERVERS_NA`, I guess

**Comment by RostakaGmfun** [comment]
Run build

File: client/client-multi/client-c/src/kaa/kaa_bootstrap_manager.c [hunk 3]
```diff
@@ -452,9 +453,10 @@ kaa_error_t kaa_bootstrap_manager_handle_server_sync(kaa_bootstrap_manager_t *se
     return error_code;
 }
 
-kaa_error_t kaa_bootstrap_manager_on_access_point_failed(kaa_bootstrap_manager_t *self
-                                                       , kaa_transport_protocol_id_t *protocol_id
-                                                       , kaa_server_type_t type)
+kaa_error_t kaa_bootstrap_manager_on_access_point_failed(kaa_bootstrap_manager_t *self,
+                                                         kaa_transport_protocol_id_t *protocol_id,
+                                                         kaa_server_type_t type,
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_bootstrap_manager.c:458]: `reason_code` now encodes error source (bootstrap / operation), so I'm not sure `type` argument is worth keeping.

File: client/client-multi/client-c/src/kaa/kaa_bootstrap_manager.h [hunk 4]
```diff
@@ -58,9 +58,10 @@ typedef enum {
  *
  * @see kaa_transport_protocol_id_t
  */
-kaa_error_t kaa_bootstrap_manager_on_access_point_failed(kaa_bootstrap_manager_t *self
-                                                       , kaa_transport_protocol_id_t *protocol_id
-                                                       , kaa_server_type_t type);
+kaa_error_t kaa_bootstrap_manager_on_access_point_failed(kaa_bootstrap_manager_t *self,
+                                                         kaa_transport_protocol_id_t *protocol_id,
+                                                         kaa_server_type_t type,
+													     uint16_t reason_code);
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_bootstrap_manager.h]: alignment

File: client/client-multi/client-c/src/kaa/platform-impl/common/ext_log_upload_strategies.c [hunk 5]
```diff
@@ -165,7 +165,7 @@ kaa_error_t ext_log_upload_strategy_on_timeout(void *context)
         KAA_RETURN_IF_ERR(error_code);
         error_code = kaa_bootstrap_manager_on_access_point_failed(self->bootstrap_manager
                                                                 , &protocol_id
-                                                                , KAA_SERVER_OPERATIONS);
+                                                                , KAA_SERVER_OPERATIONS, 0);
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/platform-impl/common/ext_log_upload_strategies.c]: `KAA_OPERATION_SERVERS_NA`

File: client/client-multi/client-c/src/kaa/platform-impl/common/kaa_failover_strategy.c [hunk 6]
```diff
@@ -49,7 +50,32 @@ void kaa_failover_strategy_destroy(kaa_failover_strategy_t* strategy)
 
 kaa_failover_decision_t kaa_failover_strategy_on_failover(void *self, kaa_failover_reason reason)
 {
-    (void)reason;
-    kaa_failover_strategy_t *strategy = (kaa_failover_strategy_t *) self;
+	 kaa_failover_strategy_t *strategy = (kaa_failover_strategy_t *) self;
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/platform-impl/common/kaa_failover_strategy.c]: 1) Indentation.
2) Cast is not needed.

**Comment by RostakaGmfun** [comment]
Run build

**Comment by RostakaGmfun** [comment]
Run build

**Comment by RostakaGmfun** [comment]
Run build

**Comment by rasendubi** [comment]
Some of comments are still not addressed

File: client/client-multi/client-c/src/kaa/kaa_protocols/kaa_tcp/kaatcp_common.h [hunk 7]
```diff
@@ -106,7 +106,8 @@ typedef enum {
     KAATCP_CONNACK_IDENTIFIER_REJECTED    = 0x03,
     KAATCP_CONNACK_SERVER_UNAVAILABLE     = 0x04,
     KAATCP_CONNACK_REFUSE_BAD_CREDENTIALS = 0x05,
-    KAATCP_CONNACK_NOT_AUTHORIZED         = 0x06
+    KAATCP_CONNACK_NOT_AUTHORIZED         = 0x06,
+    KAATCP_CONNACK_REFUSE_VERIFICATION_FAILED = 0x10,
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_protocols/kaa_tcp/kaatcp_common.h:110]: Align

File: client/client-multi/client-c/src/kaa/kaa_protocols/kaa_tcp/kaatcp_common.h [hunk 8]
```diff
@@ -117,6 +118,7 @@ typedef enum {
     KAATCP_DISCONNECT_NONE           = 0x00,
     KAATCP_DISCONNECT_BAD_REQUEST    = 0x01,
     KAATCP_DISCONNECT_INTERNAL_ERROR = 0x02,
+    KAATCP_DISCONNECT_CREDENTIALS_REVOKED = 0x03,
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_protocols/kaa_tcp/kaatcp_common.h:121]: Align

File: client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c [hunk 9]
```diff
@@ -372,7 +373,7 @@ kaa_error_t kaa_tcp_channel_sync_handler(void *context, const kaa_extension_id s
                                       , SOCKET_CONNECTION_ERROR
                                       , ((kaa_tcp_channel_t *) context)->access_point.socket_descriptor);
 
-            error_code = kaa_tcp_channel_on_access_point_failed((kaa_tcp_channel_t *) context);
+            error_code = kaa_tcp_channel_on_access_point_failed((kaa_tcp_channel_t *) context, KAA_CHANNEL_NA);
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c]: Redundant cast

File: client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c [hunk 10]
```diff
@@ -962,11 +979,13 @@ void kaa_tcp_channel_disconnect_message_callback(void *context, kaatcp_disconnec
 
     KAA_LOG_TRACE(((kaa_tcp_channel_t *) context)->logger, KAA_ERR_NONE,"Kaa TCP channel [0x%08X] DISCONNECT message received"
                                                                             , ((kaa_tcp_channel_t *) context)->access_point.id);
-
-    if (message.reason != KAATCP_DISCONNECT_INTERNAL_ERROR) {
+    if (message.reason == KAATCP_DISCONNECT_CREDENTIALS_REVOKED) {
+        kaa_tcp_channel_socket_io_error(((kaa_tcp_channel_t *) context), KAA_CREDENTIALS_REVOKED);
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c]: Redundant cast

File: client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c [hunk 11]
```diff
@@ -962,11 +979,13 @@ void kaa_tcp_channel_disconnect_message_callback(void *context, kaatcp_disconnec
...
+    }
+    if (message.reason != KAATCP_DISCONNECT_INTERNAL_ERROR && message.reason != KAATCP_DISCONNECT_CREDENTIALS_REVOKED) {
         ((kaa_tcp_channel_t *) context)->sync_state = KAA_TCP_CHANNEL_SYNC_OP_STARTED;
     }
-    kaa_tcp_channel_socket_io_error(((kaa_tcp_channel_t *) context));
+    kaa_tcp_channel_socket_io_error(((kaa_tcp_channel_t *) context), KAA_CHANNEL_NA);
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c]: Redundant cast

File: client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c [hunk 12]
```diff
@@ -927,30 +936,38 @@ void kaa_tcp_channel_connack_message_callback(void *context, kaatcp_connack_t me
...
+        } else if (message.return_code == KAATCP_CONNACK_REFUSE_VERIFICATION_FAILED) {
+            KAA_LOG_WARN(channel->logger, KAA_ERR_NONE,
+                         "Kaa TCP channel receiver KAATCP_CONNACK_REFUSE_VERIFICATION_FAILED",
+                         channel->access_point.id);
+            kaa_tcp_channel_socket_io_error(channel, KAA_ENDPOINT_NOT_REGISTERED);
         } else {
-            KAA_LOG_WARN(channel->logger, KAA_ERR_BAD_STATE, "Kaa TCP channel [0x%08X] authorization failed, code %d"
-                                                                                , channel->access_point.id, message.return_code);
-            kaa_tcp_channel_socket_io_error(channel);
+            KAA_LOG_WARN(channel->logger, KAA_ERR_BAD_STATE,
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/platform-impl/common/kaa_tcp_channel.c]: I suppose that KAA_LOG_WARN should be converted to KAA_LOG_ERROR

**Comment by forGGe** [comment]
Reviewed

**Comment by forGGe** [comment]
looks good

**Pull request merged**